flink-table/flink-table-planner-loader/src/main/java/org/apache/flink/table/planner/loader/PlannerModule.java

```diff
@@ -0,0 +1,143 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.planner.loader;
+
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.configuration.ConfigurationUtils;
+import org.apache.flink.configuration.CoreOptions;
+import org.apache.flink.core.classloading.ComponentClassLoader;
+import org.apache.flink.core.classloading.SubmoduleClassLoader;
+import org.apache.flink.runtime.rpc.RpcSystem;
+import org.apache.flink.table.api.TableException;
+import org.apache.flink.table.delegation.ExecutorFactory;
+import org.apache.flink.table.delegation.ExpressionParserFactory;
+import org.apache.flink.table.delegation.PlannerFactory;
+import org.apache.flink.table.factories.FactoryUtil;
+import org.apache.flink.util.IOUtils;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Arrays;
+import java.util.UUID;
+import java.util.stream.Stream;
+
+/**
+ * Module holder that loads the flink-table-planner module in a separate classpath.
+ *
+ * <p>This loader expects the flink-table-planner jar to be accessible via {@link
+ * ClassLoader#getResource(String)}. It will extract the jar into a temporary directory and create a
+ * new {@link SubmoduleClassLoader} to load the various planner factories from that jar.
+ */
+class PlannerModule {
+
+    /**
+     * The name of the table planner dependency jar, bundled with flink-table-planner-loader module
+     * artifact.
+     */
+    static final String FLINK_TABLE_PLANNER_FAT_JAR = "flink-table-planner.jar";
+
+    static final String HINT_USAGE =
+            "mvn clean package -pl flink-table/flink-table-planner,flink-table/flink-table-planner-loader -DskipTests";
+
+    private final ClassLoader submoduleClassLoader;
+
+    private PlannerModule() {
+        try {
+            final ClassLoader flinkClassLoader = RpcSystem.class.getClassLoader();
+
+            final Path tmpDirectory =
+                    Paths.get(ConfigurationUtils.parseTempDirectories(new Configuration())[0]);
+            Files.createDirectories(tmpDirectory);
+            final Path tempFile =
+                    Files.createFile(
+                            tmpDirectory.resolve(
+                                    "flink-table-planner_" + UUID.randomUUID() + ".jar"));
+
+            final InputStream resourceStream =
+                    flinkClassLoader.getResourceAsStream(FLINK_TABLE_PLANNER_FAT_JAR);
+            if (resourceStream == null) {
+                throw new TableException(
```

Comment:
> I guess this will also be the exception if the planner has been removed from /opt.

WDYM? This has nothing to do with `/opt`, as we bundle the planner directly in the uber jar of the flink-table-planner-loader (try to build it and unzip it)

Reply:
True, I initially thought we only package the planner into the dist once.